homeassistant/components/media_player/androidtv.py

```diff
@@ -0,0 +1,413 @@
+    "sysright": "283",
+    "sysup": "280",
+    "text": "233",
+    "top": "122",
+    "up": "19",
+    "vga": "251",
+    "voldown": "25",
+    "volup": "24",
+    "yellow": "185"
+}
+
+DATA_KEY = '{}.androidtv'.format(DOMAIN)
+
+
+def setup_platform(hass, config, add_entities, discovery_info=None):
+    """Set up the androidtv platform."""
+
+    if DATA_KEY not in hass.data:
+        hass.data[DATA_KEY] = {}
+
+    from adb.client import Client as AdbClient
+    client = AdbClient(host="127.0.0.1", port=5037)
+
+    androidtvs = []
+    for host, device_config in config[CONF_DEVICES].items():
+        name = device_config[CONF_NAME]
+        port = str(device_config[CONF_PORT])
+        scan_interval = device_config[CONF_SCAN_INTERVAL]
+        uri = host + ':' + port
+
+        try:
+            ADBdev = client.device(uri)
+            if ADBdev is None:
+                _LOGGER.error(
+                    "Can't reach adb server, is it running ?")
+                raise PlatformNotReady
+
+            device = {'name': name,
+                      'ipaddr': host,
+                      'port': port,
+                      'scan_interval': scan_interval}
+            androidtv = AndroidTv(device)
+            androidtvs.append(androidtv)
+            hass.data[DATA_KEY][host] = androidtv
+        except RuntimeError:
+            _LOGGER.error(
+                "Can't reach adb server, is it running ?")
+            raise PlatformNotReady
+
+    add_entities(androidtvs, True)
+
+    def service(service):
+        """Dispatch service calls to target entities."""
+        params = {key: value for key, value in service.data.items()
+                  if key != ATTR_ENTITY_ID}
+
+        entity_ids = service.data.get(ATTR_ENTITY_ID)
+        if entity_ids:
+            target_devices = [dev for dev in hass.data[DATA_KEY].values()
+                              if dev.entity_id in entity_ids]
+        else:
+            target_devices = hass.data[DATA_KEY].values()
+
+        for target_device in target_devices:
+            if len(params.keys()) == 1:
+                if 'action' in params:
+                    target_device.do_action(params['action'])
+                elif 'intent' in params:
+                    target_device.start_intent(params['intent'])
+                elif 'key' in params:
+                    target_device.input_key(params['key'])
+
+    hass.services.register(DOMAIN, 'androidtv_action', service)
+
+
+class AndroidTv(MediaPlayerDevice):
+    """Representation of an AndroidTv device."""
+
+    def __init__(self, device):
+        """Initialize the Android device."""
+        from adb.client import Client as AdbClient
+        host = device['ipaddr']
+        port = device['port']
+        name = device['name']
+        scan_interval = device['scan_interval']
+        self._scan_interval = scan_interval
+
+        self._uri = host + ':' + port
+        self._client = AdbClient(host="127.0.0.1", port=5037)
```

Comment:
Why instantiate a new client here when that has already been done in `setup_platform`?

Reply:
First time coding in an oriented object way, wasn't 100% sure of what was happening.